File: starid/entrypoint_main.py
"""entrypoint for pure python workflows. intended as 'hello world', bare minimum, first thing to get running. also
intended to be directly comparable with libstatid-debug/entrypoint_main.cpp"""
import os
import pickle
from starid.util.get_project_root import get_project_root
from starid.sky.sky import Sky
from starid.triangles.star_pairs import StarPairs
from starid.triangles.settler import SETTLER
from starid.triangles.nomad import NOMAD

dirdata = get_project_root() + '/data/'
dirtmp = get_project_root() + '/tmp/'
pathskymap = dirdata + "skymap.txt"
pathsky = dirtmp + "sky.pickle"
pathstarpairs = dirtmp + "starpairs.pickle"
doyaw = False
doplot = False
target = 3

def main():
    """directly comparable with main() of libstatid-debug/entrypoint_main.cp"""
    sky = sky_util(pathsky, pathskymap)
    image = sky.image_of_target(target, doyaw=doyaw)
    if doplot: image.plot()

    starpairs = starpairs_util(sky, pathstarpairs)

    settler = SETTLER(starpairs)
    result_settler = settler.run(image)

    nomad = NOMAD(starpairs)
    result_nomad = nomad.run(image)

    return

def sky_util(pathsky_, pathskymap_):
    """load sky object from disk. if not present, create and write."""
    if not os.path.exists(dirtmp): os.makedirs(dirtmp)
    try:
        sky = pickle.load(open(pathsky_, 'rb'))
    except (Exception,):
        sky = Sky()
        sky.generate(pathskymap_)
        pickle.dump(sky, open(pathsky_, 'wb'))
    return sky

def starpairs_util(sky_, pathstarpairs_):
    """load starpairs object from disk. if not present, create and write."""
    if not os.path.exists(dirtmp): os.makedirs(dirtmp)
    try:
        starpairs = pickle.load(open(pathstarpairs_, 'rb'))
    except (Exception,):
        starpairs = StarPairs()
        starpairs.generate(sky_)
        pickle.dump(starpairs, open(pathstarpairs_, 'wb'))
    return starpairs

if __name__ == '__main__':
    main()

File: libstarid-debug/entrypoint_main.cpp
#include <cereal/archives/binary.hpp>
#include <fstream>
#include "sky/sky.h"
#include "triangles/star_pairs.h"
#include "triangles/settler.h"
#include "triangles/nomad.h"
using namespace starid;
using namespace std;
static string dirdata = "../../data/";
static string dirtmp = "./";
static string pathskymap = dirdata + "skymap.txt";
static string pathsky = dirtmp + "sky.cereal";
static string pathstarpairs = dirtmp + "starpairs.cereal";
Sky sky_util(string &pathsky, string &pathskymap);
Starpairs starpairs_util(Sky &sky, string &pathstarpairs);

int main() {
    auto sky = sky_util(pathsky, pathskymap);
    auto image = sky.image_generator(3);
    auto starpairs = starpairs_util(sky, pathstarpairs);
//    NOMAD nomad(starpairs);
//    auto id2 = nomad.run(image["pixels"]);
    SETTLER settler(starpairs);
    auto id1 = settler.run(image["pixels"]);
    return 0;
}

Sky sky_util(string &pathsky_, string &pathskymap_) {
    Sky sky;
    try {
        ifstream is1(pathsky_, ios::binary);
        cereal::BinaryInputArchive iarchive1(is1);
        iarchive1(sky);
    } catch (...) {
        sky.generate(pathskymap_);
        ofstream os1(pathsky_, ios::binary);
        cereal::BinaryOutputArchive oarchive1(os1);
        oarchive1(sky);
    }
    return sky;
}

Starpairs starpairs_util(Sky &sky, string &pathstarpairs_) {
    Starpairs starpairs;
    try {
        ifstream is2(pathstarpairs_, ios::binary);
        cereal::BinaryInputArchive iarchive2(is2);
        iarchive2(starpairs);
    } catch (...) {
        starpairs.generate(sky);
        ofstream os2(pathstarpairs_, ios::binary);
        cereal::BinaryOutputArchive oarchive2(os2);
        oarchive2(starpairs);
    }
    return starpairs;
}
